test_conformance/subgroups/test_workgroup.cpp

```diff
@@ -1648,584 +1549,49 @@ struct RED_NU_LG {
             y += nw;
             m += 4 * nw;
         }
+
         return 0;
     }
 };
 
-// DESCRIPTION:
-// Test for reduce cluster functions
-// Which: 0 - add, 1 - max, 2 - min, 3 - mul, 4 - and, 5 - or, 6 - xor
+// Scan Inclusive functions
 template <typename Ty, int Which>
-struct RED_CLU {
-    static void gen(Ty *x, Ty *t, cl_int *m, int ns, int nw, int ng, int a=0)
+struct SCIN {
+    static void gen(Ty *x, Ty *t, cl_int *m, int ns, int nw, int ng)
     {
-        int i, ii, j, k, n, cluster_index;
-        int nj = (nw + ns - 1) / ns;
+        genrand<Ty, Which>(x, t, m, ns, nw, ng);
+    }
 
-        ii = 0;
-        std::vector<unsigned int> cluster_values;
-        for (int c = 2; c <= ns; c = c << 1) {
-            cluster_values.push_back(c);
-        }
+    static int chk(Ty *x, Ty *y, Ty *mx, Ty *my, cl_int *m, int ns, int nw, int ng)
+    {
+        int ii, i, j, k, n;
+        int nj = (nw + ns - 1)/ns;
+        Ty tr, rr;
 
-        for (k = 0; k < ng; ++k) {
-            for (j = 0; j < nj; ++j) {
-                ii = j * ns;
+        log_info("  sub_group_scan_inclusive_%s(%s)...\n",  operation_names[Which], TypeName<Ty>::val());
+
+        for (k=0; k<ng; ++k) {
+            // Map to array indexed to array indexed by local ID and sub group
+            for (j=0; j<nw; ++j) {
+                i = m[4*j+1]*ns + m[4*j];
+                mx[i] = x[j];
+                my[i] = y[j];
+            }
+
+            for (j=0; j<nj; ++j) {
+                ii = j*ns;
                 n = ii + ns > nw ? nw - ii : ns;
 
-                cluster_index = (int)(genrand_int32(gMTdata) & 0x7fffffff) % cluster_values.size(); //randomize cluster size
+                // Check result
+                for (i=0; i<n; ++i) {
+                    tr = i == 0 ? mx[ii] : OPERATION<Ty, Which>::calculate(tr, mx[ii + i]);
 
-                for (i = 0; i < n; ++i) {
-                    int midx = 4 * ii + 4 * i + 2;
-                    m[midx] = cluster_values[cluster_index];           // the third index - store information about cluster size
```

Comment:
I wanted to test different values that is stored in m[midx]. Your changes cause that only one const value will be tested == CLUSTER_SIZE

Reply:
The [spec](https://gitlab.khronos.org/opencl/OpenCL-Docs/blob/rename-clustered-reductions/ext/cl_khr_subgroup_extensions.asciidoc) says:

> The specified cluster size must be an integer **constant** expression that is a power-of-two.

(emphasis mine)

The tested functions crash if `clustersize` is not constant.

Reply:
Having said that, I agree we do want to test different cluster sizes. But this must be done in a way that cluster size appears as constant to the kernel.

Reply:
Maybe this is difficult to modify current implementation to not brake some rules but...
Could I give as kernel argument array witch set of const values ?

Reply:
What breaks is IGC casting `clustersize` as an LLVM constant. So a constant buffer won't work.

I suggest starting with my code (because it works) and adding tests for different cluster sizes later.